File: sdk/typescript/src/server/analytics.ts
import type { HttpTransport } from "../http.js";

/**
 * 事件分析面（docs/design/analytics.md §4.1）：服务端权威事件摄入 +
 * 固定形状查询。项目上下文来自凭证（API key 绑定项目；admin 会话带
 * X-Torchwood-Project）。时间参数一律 RFC3339 字符串；日粒度 UTC 切日。
 */

/** 单个服务端事件（可信代报 user_id；上限：批 100 / 名 64 / 键 25 / 16KiB）。 */
export interface ServerAnalyticsEvent {
  name: string;
  /** RFC3339；缺省 = 服务端 now（服务端钳制 [now-24h, now+5min]）。 */
  occurred_at?: string;
  props?: Record<string, unknown>;
  session_id?: string;
  /** 可信代报的事件归属用户；缺省 = 无归属（仅进总量口径）。 */
  user_id?: string;
}

export interface IngestEventsResult {
  accepted: number;
  skipped: number;
}

export interface AnalyticsOverviewKpi {
  total_events: string;
  unique_users: string;
  new_users: string;
  events_per_user: number;
}

export interface AnalyticsTopEvent {
  name: string;
  total: string;
}

export interface AnalyticsTodayStats {
  total_events: string;
  unique_users: string;
}

export interface AnalyticsOverview {
  kpi?: AnalyticsOverviewKpi;
  top_events?: AnalyticsTopEvent[];
  today?: AnalyticsTodayStats;
  /** 窗口 KPI 口径：rollup | raw（worker 未覆盖时回退 raw）。 */
  source?: string;
}

export interface AnalyticsEventDefinition {
  name: string;
  first_seen?: string;
  last_seen?: string;
  total_30d?: string;
}

export interface ListEventDefinitionsResult {
  definitions?: AnalyticsEventDefinition[];
  meta?: {
    total_count?: string;
    page_size?: number;
    next_page_token?: string;
    prev_page_token?: string;
  };
}

export interface AnalyticsTimeseriesPoint {
  bucket?: string;
  total?: string;
  unique_users?: string;
}

export interface QueryTimeseriesResult {
  points?: AnalyticsTimeseriesPoint[];
  /** rollup | raw（HOUR 恒 raw，DAY 优先 rollup）。 */
  source?: string;
}

export interface AnalyticsBreakdownBucket {
  value?: string;
  total?: string;
  unique_users?: string;
}

export interface QueryBreakdownResult {
  buckets?: AnalyticsBreakdownBucket[];
  /** breakdown v1 恒 raw。 */
  source?: string;
}

export interface AnalyticsRetentionCohort {
  cohort?: string;
  size?: string;
  /** retained[k] = cohort+k 日仍活跃用户数（D0–D14）。 */
  retained?: string[];
}

export interface QueryRetentionResult {
  cohorts?: AnalyticsRetentionCohort[];
  /** retention 恒 rollup（user_days / first_seen 基座）。 */
  source?: string;
}

export interface AnalyticsUserEvent {
  id?: string;
  name?: string;
  occurred_at?: string;
  ingested_at?: string;
  /** 摄入通道 client | server。 */
  source?: string;
  platform?: string;
  app_version?: string;
  session_id?: string;
  props?: Record<string, unknown>;
}

export interface ListUserEventsResult {
  events?: AnalyticsUserEvent[];
  meta?: {
    total_count?: string;
    page_size?: number;
    next_page_token?: string;
    prev_page_token?: string;
  };
  /** 下钻恒 raw。 */
  source?: string;
}

export class AnalyticsService {
  constructor(private readonly http: HttpTransport) {}

  /**
   * 服务端权威事件批量摄入（POST /v1/server/analytics/events）。
   * 部分接收语义：坏事件计入 skipped；at-least-once 不做去重。
   */
  async ingest(events: ServerAnalyticsEvent[]): Promise<IngestEventsResult> {
    return this.http.request<IngestEventsResult>("POST", "/v1/server/analytics/events", {
      auth: "apiKey",
      body: { events },
    });
  }

  /** 窗口 KPI + Top 事件 + 今日实时数（GET /v1/server/analytics/overview）。 */
  async getOverview(periodStart: string, periodEnd: string): Promise<AnalyticsOverview> {
    return this.http.request<AnalyticsOverview>("GET", "/v1/server/analytics/overview", {
      auth: "apiKey",
      query: { period_start: periodStart, period_end: periodEnd },
    });
  }

  /** 事件字典分页（GET /v1/server/analytics/event-definitions）。 */
  async listEventDefinitions(params?: { page_size?: number; page_token?: string }): Promise<AnalyticsEventDefinition[]> {
    const res = await this.listEventDefinitionWithMeta(params);
    return res.definitions ?? [];
  }

  /** 同 listEventDefinitions，但携带分页 meta。 */
  async listEventDefinitionWithMeta(params?: {
    page_size?: number;
    page_token?: string;
  }): Promise<ListEventDefinitionsResult> {
    return this.http.request<ListEventDefinitionsResult>("GET", "/v1/server/analytics/event-definitions", {
      auth: "apiKey",
      query: { page_size: params?.page_size, page_token: params?.page_token },
    });
  }

  /** 事件趋势（GET /v1/server/analytics/timeseries；HOUR 走 raw、DAY 走 rollup）。 */
  async queryTimeseries(params: {
    period_start: string;
    period_end: string;
    granularity: "HOUR" | "DAY";
    /** 事件名集合（≤10）；缺省 = 全部事件聚合。 */
    names?: string[];
  }): Promise<QueryTimeseriesResult> {
    return this.http.request<QueryTimeseriesResult>("GET", "/v1/server/analytics/timeseries", {
      auth: "apiKey",
      query: {
        names: params.names,
        period_start: params.period_start,
        period_end: params.period_end,
        granularity: params.granularity,
      },
    });
  }

  /**
   * 维度拆解（GET /v1/server/analytics/breakdown；Top-N + __other__ 归并，
   * 窗口护栏 ≤30 天）。prop_key 服务端白名单校验后参数化。
   */
  async queryBreakdown(params: {
    name: string;
    prop_key: string;
    period_start: string;
    period_end: string;
    top_n?: number;
  }): Promise<QueryBreakdownResult> {
    return this.http.request<QueryBreakdownResult>("GET", "/v1/server/analytics/breakdown", {
      auth: "apiKey",
      query: {
        name: params.name,
        prop_key: params.prop_key,
        period_start: params.period_start,
        period_end: params.period_end,
        top_n: params.top_n,
      },
    });
  }

  /** 留存矩阵（GET /v1/server/analytics/retention；cohort 窗 ≤92 天 × D0–D14）。 */
  async queryRetention(params: { cohort_start: string; cohort_end: string }): Promise<QueryRetentionResult> {
    return this.http.request<QueryRetentionResult>("GET", "/v1/server/analytics/retention", {
      auth: "apiKey",
      query: { cohort_start: params.cohort_start, cohort_end: params.cohort_end },
    });
  }

  /**
   * 用户行为轨迹下钻（GET /v1/server/analytics/users/{user_id}/events；
   * keyset 分页，可选时间窗 ≤92 天）。
   */
  async listUserEvents(
    userId: string,
    params?: { page_size?: number; page_token?: string; period_start?: string; period_end?: string }
  ): Promise<AnalyticsUserEvent[]> {
    const res = await this.listUserEventsWithMeta(userId, params);
    return res.events ?? [];
  }

  /** 同 listUserEvents，但携带分页 meta 与 source 口径。 */
  async listUserEventsWithMeta(
    userId: string,
    params?: { page_size?: number; page_token?: string; period_start?: string; period_end?: string }
  ): Promise<ListUserEventsResult> {
    return this.http.request<ListUserEventsResult>(
      "GET",
      `/v1/server/analytics/users/${encodeURIComponent(userId)}/events`,
      {
        auth: "apiKey",
        query: {
          page_size: params?.page_size,
          page_token: params?.page_token,
          period_start: params?.period_start,
          period_end: params?.period_end,
        },
      }
    );
  }
}

File: cli/analytics.go
package cli

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"os"

	"github.com/lynx-go/commands"
)

const (
	methodAnalyticsIngest      = "/torchwood.server.v1.AnalyticsService/IngestEvents"
	methodAnalyticsOverview    = "/torchwood.server.v1.AnalyticsService/GetOverview"
	methodAnalyticsDefinitions = "/torchwood.server.v1.AnalyticsService/ListEventDefinitions"
	methodAnalyticsTimeseries  = "/torchwood.server.v1.AnalyticsService/QueryTimeseries"
	methodAnalyticsBreakdown   = "/torchwood.server.v1.AnalyticsService/QueryBreakdown"
	methodAnalyticsRetention   = "/torchwood.server.v1.AnalyticsService/QueryRetention"
	methodAnalyticsUserEvents  = "/torchwood.server.v1.AnalyticsService/ListUserEvents"
)

// newAnalyticsCmd 覆盖 AnalyticsService 全部 7 个方法：服务端权威事件摄入
// （ingest，analytics.write）与固定形状查询（overview/events/timeseries/
// breakdown/retention/user-events，analytics.read）。窗口护栏（HOUR ≤7 天 /
// DAY ≤366 天 / breakdown ≤30 天 / retention cohort ≤92 天 / user-events
// ≤92 天）由服务端执行。
func newAnalyticsCmd(g *GlobalFlags) *group {
	return newGroup(g, "analytics", "event analytics: server-side ingest (analytics.write) and fixed-shape queries (analytics.read)", func(sub *commands.App) {
		sub.Register(
			newAnalyticsIngestCmd(g),
			newAnalyticsOverviewCmd(g),
			newAnalyticsEventsCmd(g),
			newAnalyticsTimeseriesCmd(g),
			newAnalyticsBreakdownCmd(g),
			newAnalyticsRetentionCmd(g),
			newAnalyticsUserEventsCmd(g),
		)
	})
}

func newAnalyticsIngestCmd(g *GlobalFlags) *verb {
	var file string
	return newVerb(g, "ingest", "ingest a batch of server-authoritative events (analytics.write; 1-100 per batch, bad events are skipped not rejected)", "analytics ingest --file <path | ->",
		func(fs *flag.FlagSet) {
			fs.StringVar(&file, "file", "", `JSON payload: a bare event array or a {"events": [...]} object ('-' reads stdin); event fields (camelCase): name (required), occurredAt (RFC3339), props, sessionId, userId`)
		},
		func(v *verb, env *commands.Environment, args []string) error {
			if err := noArgs(v, args); err != nil {
				return err
			}
			if file == "" {
				return fmt.Errorf("--file is required (path or '-' for stdin)")
			}
			req, err := buildIngestReq(file)
			if err != nil {
				return err
			}
			return call(g, env, methodAnalyticsIngest, req)
		})
}

// buildIngestReq 读入 --file 载荷并构造 IngestServerEventsRequest。
func buildIngestReq(path string) (map[string]any, error) {
	var content []byte
	var err error
	if path == "-" {
		content, err = io.ReadAll(os.Stdin)
	} else {
		content, err = os.ReadFile(path) // #nosec G304 -- path from --file flag
	}
	if err != nil {
		return nil, fmt.Errorf("failed to read --file: %v", err)
	}
	return ingestEventsPayload(content)
}

// ingestEventsPayload 解析载荷：裸事件数组包一层 events 键；对象原样透传
// （须含 events）。形状与批量上限校验归服务端 protovalidate。JSONL（一行一
// 事件）会在这里被静默截断成首个事件，必须显式拒绝。
func ingestEventsPayload(content []byte) (map[string]any, error) {
	dec := json.NewDecoder(bytes.NewReader(content))
	dec.UseNumber()
	var v any
	if err := dec.Decode(&v); err != nil {
		return nil, fmt.Errorf("failed to parse ingest payload: %v", err)
	}
	if dec.More() {
		return nil, fmt.Errorf(`ingest payload must be a single JSON array or object (one event per line is not supported; wrap the events in an array)`)
	}
	switch t := v.(type) {
	case []any:
		return map[string]any{"events": t}, nil
	case map[string]any:
		return t, nil
	default:
		return nil, fmt.Errorf(`ingest payload must be a JSON array of events or an {"events": [...]} object`)
	}
}

func newAnalyticsOverviewCmd(g *GlobalFlags) *verb {
	var from, to string
	return newVerb(g, "overview", "window KPIs + top events + today's real-time counts (source: rollup|raw)", "analytics overview --from <ts> --to <ts>",
		func(fs *flag.FlagSet) {
			registerWindowFlags(fs, &from, &to)
		},
		func(v *verb, env *commands.Environment, args []string) error {
			if err := noArgs(v, args); err != nil {
				return err
			}
			if err := requireWindow(from, to); err != nil {
				return err
			}
			return call(g, env, methodAnalyticsOverview, map[string]any{"periodStart": tsJSON(from), "periodEnd": tsJSON(to)})
		})
}

// newAnalyticsEventsCmd: analytics events list —— 事件字典（自由上报 +
// 事后发现）。
func newAnalyticsEventsCmd(g *GlobalFlags) *group {
	return newGroup(g, "events", "event dictionary (definitions discovered from ingestion)", func(sub *commands.App) {
		sub.Register(newAnalyticsEventsListCmd(g))
	})
}

func newAnalyticsEventsListCmd(g *GlobalFlags) *verb {
	var pageSize int
	var pageToken string
	return newVerb(g, "list", "list event definitions (name, first/last seen, 30d total)", "analytics events list [--page-size <n>] [--page-token <t>]",
		func(fs *flag.FlagSet) {
			fs.IntVar(&pageSize, "page-size", 0, "page size (0-500, server default when omitted)")
			fs.StringVar(&pageToken, "page-token", "", "next page token from the previous response")
		},
		func(v *verb, env *commands.Environment, args []string) error {
			if err := noArgs(v, args); err != nil {
				return err
			}
			return call(g, env, methodAnalyticsDefinitions, listJSON(pageSize, pageToken))
		})
}

func newAnalyticsTimeseriesCmd(g *GlobalFlags) *verb {
	var names, from, to, granularity string
	return newVerb(g, "timeseries", "event counts over time (hour → raw, day → rollup)", "analytics timeseries [--names '<json array>'] --from <ts> --to <ts> --granularity <hour|day>",
		func(fs *flag.FlagSet) {
			fs.StringVar(&names, "names", "", `event names as a JSON array (up to 10); omit for all events aggregated`)
			registerWindowFlags(fs, &from, &to)
			fs.StringVar(&granularity, "granularity", "", "bucket size: hour | day (required)")
		},
		func(v *verb, env *commands.Environment, args []string) error {
			if err := noArgs(v, args); err != nil {
				return err
			}
			if err := requireWindow(from, to); err != nil {
				return err
			}
			gran, err := granularityJSON(granularity)
			if err != nil {
				return err
			}
			req := map[string]any{"periodStart": tsJSON(from), "periodEnd": tsJSON(to), "granularity": gran}
			if names != "" {
				list, err := jsonStringList(names, "--names")
				if err != nil {
					return err
				}
				req["names"] = list
			}
			return call(g, env, methodAnalyticsTimeseries, req)
		})
}

// granularityJSON 把 hour|day 映射为 protojson 接受的枚举名（全名原样透传）。
func granularityJSON(s string) (string, error) {
	switch s {
	case "hour":
		return "ANALYTICS_GRANULARITY_HOUR", nil
	case "day":
		return "ANALYTICS_GRANULARITY_DAY", nil
	case "ANALYTICS_GRANULARITY_HOUR", "ANALYTICS_GRANULARITY_DAY":
		return s, nil
	default:
		return "", fmt.Errorf("invalid --granularity %q: use hour or day", s)
	}
}

func newAnalyticsBreakdownCmd(g *GlobalFlags) *verb {
	var name, propKey, from, to string
	var topN int
	return newVerb(g, "breakdown", "break an event down by a property key (raw, top-N + __other__, window ≤30 days)", "analytics breakdown --name <event> --prop-key <key> --from <ts> --to <ts> [--top-n <n>]",
		func(fs *flag.FlagSet) {
			fs.StringVar(&name, "name", "", "event name (required)")
			fs.StringVar(&propKey, "prop-key", "", "property key to group by (validated server-side against an allowlist to prevent injection)")
			registerWindowFlags(fs, &from, &to)
			fs.IntVar(&topN, "top-n", 0, "top-N buckets (0-50, remaining merged into __other__; default 20)")
		},
		func(v *verb, env *commands.Environment, args []string) error {
			if err := noArgs(v, args); err != nil {
				return err
			}
			if name == "" || propKey == "" {
				return fmt.Errorf("--name and --prop-key are required")
			}
			if err := requireWindow(from, to); err != nil {
				return err
			}
			req := map[string]any{"name": name, "propKey": propKey, "periodStart": tsJSON(from), "periodEnd": tsJSON(to)}
			if topN > 0 {
				req["topN"] = topN
			}
			return call(g, env, methodAnalyticsBreakdown, req)
		})
}

func newAnalyticsRetentionCmd(g *GlobalFlags) *verb {
	var from, to string
	return newVerb(g, "retention", "retention matrix (cohort × D0-D14, day granularity UTC; cohort window ≤92 days)", "analytics retention --from <ts> --to <ts>",
		func(fs *flag.FlagSet) {
			registerWindowFlags(fs, &from, &to)
		},
		func(v *verb, env *commands.Environment, args []string) error {
			if err := noArgs(v, args); err != nil {
				return err
			}
			if err := requireWindow(from, to); err != nil {
				return err
			}
			return call(g, env, methodAnalyticsRetention, map[string]any{"cohortStart": tsJSON(from), "cohortEnd": tsJSON(to)})
		})
}

func newAnalyticsUserEventsCmd(g *GlobalFlags) *verb {
	var from, to, pageToken string
	var pageSize int
	return newVerb(g, "user-events", "drill into one user's event trail (raw, newest first, keyset pagination, window ≤92 days)", "analytics user-events <user-id> [--page-size <n>] [--page-token <t>] [--from <ts>] [--to <ts>]",
		func(fs *flag.FlagSet) {
			fs.IntVar(&pageSize, "page-size", 0, "page size (0-100, server default when omitted)")
			fs.StringVar(&pageToken, "page-token", "", "opaque cursor from the previous page")
			fs.StringVar(&from, "from", "", "optional window start (RFC3339 or YYYY-MM-DD)")
			fs.StringVar(&to, "to", "", "optional window end")
		},
		func(v *verb, env *commands.Environment, args []string) error {
			if err := exactArgs(v, args, 1); err != nil {
				return err
			}
			req := map[string]any{"userId": args[0]}
			if pageSize > 0 {
				req["pageSize"] = pageSize
			}
			if pageToken != "" {
				req["pageToken"] = pageToken
			}
			if from != "" {
				req["periodStart"] = tsJSON(from)
			}
			if to != "" {
				req["periodEnd"] = tsJSON(to)
			}
			return call(g, env, methodAnalyticsUserEvents, req)
		})
}

// registerWindowFlags 挂 --from/--to（分析查询的公共窗口旗标）。
func registerWindowFlags(fs *flag.FlagSet, from, to *string) {
	fs.StringVar(from, "from", "", "window start (RFC3339, or YYYY-MM-DD = UTC midnight)")
	fs.StringVar(to, "to", "", "window end (same formats)")
}

// requireWindow 校验必填时间窗。
func requireWindow(from, to string) error {
	if from == "" || to == "" {
		return fmt.Errorf("--from and --to are required (RFC3339, or YYYY-MM-DD = UTC midnight)")
	}
	return nil
}

// tsJSON 归一时间旗标：纯日期（YYYY-MM-DD）补为 UTC 零点 RFC3339，其余原样
// 透传（格式合法性由 protojson/服务端校验）。
func tsJSON(s string) string {
	if len(s) == 10 && s[4] == '-' && s[7] == '-' {
		return s + "T00:00:00Z"
	}
	return s
}
